cloud-sql/postgres/sqlalchemy/main.py

```diff
@@ -204,30 +57,65 @@ def create_tables():
         )
 
 
-@app.route('/', methods=['GET'])
-def index():
-    context = get_index_context()
-    return render_template('index.html', **context)
+# This global variable is declared with a value of `None`, instead of calling
+# `init_connection_pool()` immediately, to simplify testing. In general, it
+# is safe to initialize your database connection pool when your script starts
+# -- there is no need to wait for the first request.
+db = None
+
+
+# init_db lazily instantiates a database connection pool. Users of Cloud Run or
+# Cloud Functions may wish to skip this lazy instantiation and connect as soon
+# as the function is loaded. This is primarily to help testing.
+@app.before_first_request
+def init_db():
+    global db
+    if db is None:
```

Comment:
Is this thread safe?

Reply:
Good catch, this should be fine for all cases other than Cloud Functions which is manually calling `init_db`.

 @kurtisvg do you have any ideas for how to handle this since each Cloud Functions request calls `init_db`?

Reply:
This almost worked, but this gets hit when the tests import the module which is prior to `app.testing` being set to True which was causing tests to fail. However this is thread-safe now since functions implementation was moved to separate file.